### PR TITLE
[AIRFLOW-361] Default handler for failures and retries in qubole operator

### PR DESCRIPTION
**Pull request by msumit**
Dear Airflow Maintainers,

Please accept this PR that addresses the following issues:
- _https://issues.apache.org/jira/browse/AIRFLOW-361_

Thanks,
Sumit

### CODE REVIEW

File: airflow/models.py [hunk 1]
```diff
@@ -336,7 +336,8 @@ def kill_zombies(self, session):
                 if ti.task_id in dag.task_ids:
                     task = dag.get_task(ti.task_id)
                     ti.task = task
-                    ti.handle_failure("{} killed as zombie".format(ti))
+                    ti.test_mode = False
```
Comment (msumit) [inline airflow/models.py]: without this line getting error `{jobs.py:793} ERROR - 'TaskInstance' object has no attribute 'test_mode'}}`

File: airflow/contrib/operators/qubole_operator.py [hunk 2]
```diff
@@ -137,7 +149,4 @@ def __setattr__(self, name, value):
         if name in QuboleOperator.template_fields:
             self.kwargs[name] = value
         else:
-            object.__setattr__(self, name, value)
-
-
-
+            object.__setattr__(self, name, value)
```
Comment (skudriashev) [inline airflow/contrib/operators/qubole_operator.py]: Please, add new line at the end.

File: airflow/contrib/hooks/qubole_hook.py [hunk 3]
```diff
@@ -84,14 +85,29 @@ def execute(self, context):
...
+
+        if cmd_id is not None:
+            cmd = self.cls.find(cmd_id)
+
+            if cmd is not None:
+                if cmd.status == 'done':
+                    logging.info('Command ID: {0} has been succeeded, hence making this TI as Success.'.format(str(cmd_id)))
+                    ti.state = State.SUCCESS
+                elif cmd.status == 'running':
+                    logging.info('Cancelling the Qubole Command Id: {0}'.format(str(cmd_id)))
```
Comment (skudriashev) [inline airflow/contrib/hooks/qubole_hook.py]: Conversion to `str` is not needed.

**Comment by skudriashev** [comment]
Please, follow 50/72 rule for commit messages.

**Comment by skudriashev** [comment]
Also, you have spelling error in the commit message: failurs -> failures.

**Comment by msumit** [comment]
@skudriashev Fixed commit message and pep8 warnings. Travis build is failing but thats not due to changes of this PR.

**Comment by msumit** [comment]
@skudriashev @criccomini @mistercrunch @bolkedebruin please review. Most of the code change is for resolving pep8 warnings.

**Comment by msumit** [comment]
Guys, any update?

File: airflow/contrib/hooks/qubole_hook.py [hunk 4]
```diff
@@ -91,20 +121,20 @@ def kill(self, ti):
...
     def get_results(self, ti=None, fp=None, inline=True, delim=None, fetch=True):
         """
-        Get results (or just s3 locations) of a command from Qubole and save it into a file
+        Get results (or just s3 locations) of a command from Qubole and save into a file
         :param ti: Task Instance of the dag, used to determine the Quboles command id
         :param fp: Optional file pointer, will create one and return if None passed
         :param inline: True to download actual results, False to get s3 locations only
-        :param delim: Replaces the CTL-A chars with the given delim, defalt to ',' if None given
-        :param fetch: when inline is True, get results directly from s3 if results are large
+        :param delim: Replaces the CTL-A chars with the given delim, defalts to ','
```
Comment (skudriashev) [inline airflow/contrib/hooks/qubole_hook.py]: defa**u**lts?

Comment (msumit) [inline airflow/contrib/hooks/qubole_hook.py]: Corrected, thanks.

**Comment by msumit** [comment]
Did a rebase to squash all commits.

**Comment by msumit** [comment]
Not sure how but after rebase landscape started showing changes which are not part of this PR. Before rebase score was +0.25%.

**Comment by msumit** [comment]
guys, plz review this.

**Comment by msumit** [comment]
@skudriashev @criccomini @mistercrunch @bolkedebruin plz review.

File: airflow/models.py (hunk 1 above)
Comment (r39132) [inline airflow/models.py]: I see this getting set in the TaskInstance models`__init__` method, which would seem the more appropriate location than here.

Comment (r39132) [inline airflow/models.py]: Can you provide a stack. Is this change needed as part of this PR or can you break this work into 2 PRs, with one PR just for this your changes to kill_zombies?

I have no issue with the other parts of your PR.

Comment (msumit) [inline airflow/models.py]: yeah, I too found it strange as in `models.py` we are setting `test_mode` to `False`. That's why I added it in PR and left a note, so someone may help in understand this behaviour. Below is the stacktrace of scheduler's `kill_zombies` function:

> [2016-08-11 12:48:19,313] {models.py:313} INFO - Finding 'running' jobs without a recent heartbeat
> [2016-08-11 12:48:19,314] {models.py:319} INFO - Failing jobs without heartbeat after 2016-08-11 12:46:04.314176
> [2016-08-11 12:48:19,321] {jobs.py:793} ERROR - 'TaskInstance' object has no attribute 'test_mode'
> Traceback (most recent call last):
>   File "/Users/sumit/src/airflow-sumit/airflow/jobs.py", line 791, in _execute
>     dagbag.kill_zombies()
>   File "/Users/sumit/src/airflow-sumit/airflow/utils/db.py", line 53, in wrapper
>     result = func(_args, *_kwargs)
>   File "/Users/sumit/src/airflow-sumit/airflow/models.py", line 341, in kill_zombies
>     False, ti.get_template_context())
>   File "/Users/sumit/src/airflow-sumit/airflow/utils/db.py", line 53, in wrapper
>     result = func(_args, *_kwargs)
>   File "/Users/sumit/src/airflow-sumit/airflow/models.py", line 1516, in get_template_context
>     'test_mode': self.test_mode,
> AttributeError: 'TaskInstance' object has no attribute 'test_mode'

Comment (msumit) [inline airflow/models.py]: OK, found the issue, actually `TaskInstance` `init` is not being called.. as explained here http://stackoverflow.com/questions/17256604/python-class-init-not-working

Comment (msumit) [inline airflow/models.py]: @r39132 still do you think that I should revert the changes from `models.py`? As in existing code as well we are doing same (line 338). Also without sending context, handling zombie is just like marking them as `failed`.

Comment (msumit) [inline airflow/models.py]: One more possible way is to overrider `__getattribute__` in `TaskInstance` class to return `False` if it's not set.

**Comment by msumit** [comment]
@r39132 limited the changes to contrib only.

**Comment by bolkedebruin** [comment]
Commit message is still incorrect needs to be in [AIRFLOW-xxx] format.

**Comment by bolkedebruin** [comment]
Also please add a description to your commit message per guidelines. Thanks!

**Comment by msumit** [comment]
@bolkedebruin @r39132 did the changes as suggested. There is some issue with landscape, it shows arbitrary errors after rebase.

**Comment by r39132** [comment]
👍 The commit message should be in active voice such as "Add default failure handler for qubole operator" and should be within the 50 characters for the commit subject. It's unclear if the 50 character limit includes the `[AIRFLOW-XXX]` prefix, but generally try to stay in that limit. I'll modify your commit message when I merge, but please be aware of it in the future.

**Comment by r39132** [comment]
Also, just noticed you didn't leave a line between the subject and body. These are the 7 rules we follow: http://chris.beams.io/posts/git-commit/#seven-rules
